### PR TITLE
Add support for JSON-RPC PubSub subscriptions

### PR DESCRIPTION
**Pull request by jvff**
Checklist for a PR:

* [x] Describe the change in **`CHANGELOG.md`**. Only applicable if the change has any impact for a user. **Internal feature, not visible to the user.**

---
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)

### CODE REVIEW

**Comment by jvff** [comment]
This PR changes the Talpid IPC client to support PubSub RPC notifications. The idea is to also update the Mullvad RPC client to use the PubSub notifications, and then use them to test for state changes in the integration tests.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-:-LBuaFOnBxDHso6AlvH9:b-93et7n)*

**Comment by jvff** [comment]
*[talpid-ipc/src/client.rs, line 56 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LBuaqL06zjSDwuecdVm:-LBuaqL06zjSDwuecdVn:bvq1py0) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2ac4639bfd1de369543c75f7b28f03499466b7d8/talpid-ipc/src/client.rs#L56)):*
> ```Rust
> }
> 
> type SubscriptionHandler = Box<Fn(JsonValue) -> SubscriptionHandlerResult + Send>;
> ```

This callback function is used to convert the `JsonValue` into the type required for the listener. It returns a special result type to indicate if it finished or not. If it returns that it has finished, it will be removed from the subscription listener list and won't receive any more events.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by jvff** [comment]
*[talpid-ipc/src/client.rs, line 401 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LBubiWvD3UIDUcbnhwi:-LBubiWxBrofxUsHo0Vf:b-9omz2r) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2ac4639bfd1de369543c75f7b28f03499466b7d8/talpid-ipc/src/client.rs#L401)):*
> ```Rust
> }
> 
> fn forward_subscription_event<V, M>(
> ```

This generic function is instanced as the listener callback for each different event type (`V` is the deserialized value, which then gets converted to the `M` message type). If deserialization fails or if the other endpoint of the `mpsc` channel is closed, it will return an error and the closure that calls this function will notify the subscription handler that it shouldn't receive any more events.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by faern** [comment]
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[talpid-ipc/src/client.rs, line 56 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LBuaqL06zjSDwuecdVm:-LBugWZc41BKCE-NKjlI:b-sqvszw) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2ac4639bfd1de369543c75f7b28f03499466b7d8/talpid-ipc/src/client.rs#L56)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

This callback function is used to convert the `JsonValue` into the type required for the listener. It returns a special result type to indicate if it finished or not. If it returns that it has finished, it will be removed from the subscription listener list and won't receive any more events.
</blockquote></details>

Is the finishing of a subscription really determined by the event callback? Is a subscription not stopped by calling unsubscribe on the server?

---

*[talpid-ipc/src/client.rs, line 147 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LBucyGDC8KHzQJenhZr:-LBucyGDC8KHzQJenhZs:b-efial1) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2ac4639bfd1de369543c75f7b28f03499466b7d8/talpid-ipc/src/client.rs#L147)):*
> ```Rust
>             }),
>             None => Ok(None),
>             _ => Err(ErrorKind::InvalidJsonRpcResponse("Invalid request ID value").into()),
> ```

I was confused about this last `_` case. I thought Hmm, but you already covered both `Option` variants. I think it might be easier to parse if this error branch was put before `None` and had the pattern matching `Some(_)` instead of just `_` to make it clear that any other `Some` value is an error.

---

*[talpid-ipc/src/client.rs, line 319 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LBuexIq4UriYoGGIEy2:-LBuexIq4UriYoGGIEy3:bodwaie) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2ac4639bfd1de369543c75f7b28f03499466b7d8/talpid-ipc/src/client.rs#L319)):*
> ```Rust
>         let raw_subscription_id = self.call(&format!("{}_subscribe", event), &[] as &[u8; 0])?;
>         let subscription_id = SubscriptionId::parse_value(&raw_subscription_id)
>             .ok_or_else(|| format!("Invalid subscription ID received: {}", raw_subscription_id))?;
> ```

You have defined a number of new error types, but this place uses raw `ErrorKind::Msg`. In libraries I think it's good if one can stick to "real" error variants. Best would be if some errors could be consolidated under the same variant as to not lead to an explosion of error variants, the chain usually tells the underlying reason anyway.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by jvff** [comment]
Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[talpid-ipc/src/client.rs, line 56 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LBuaqL06zjSDwuecdVm:-LBw-eSv018t3oBx2HRb:b-kgh052) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2ac4639bfd1de369543c75f7b28f03499466b7d8/talpid-ipc/src/client.rs#L56)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Is the finishing of a subscription really determined by the event callback? Is a subscription not stopped by calling unsubscribe on the server?
</blockquote></details>

I updated the implementation to have an `unsubscribe` method and I updated the subscription event callback closure to call the `unsubscribe` method when processing the event fails.

However, this was not a simple change. Because the closure runs on a different thread, it was necessary to wrap all of the fields in `Arc<Mutex<_>>`. Then, the problem was that if I simply shared the references between the "main" thread and the "handler" thread, there was the risk of the connection not being closed when all of the user references were dropped. So I decided to make a "weak reference" `WeakWsIpcClient` that works around that limitation by only unsubscribing if the references are still valid.

Things got more complex than expected. I'm not sure the current solution is the best one. However, simply removing all this and letting the user be responsible for unsubscribing seems like a missed opportunity.

I can think of another alternative. I'll see if it works and I'll post it as a new revision if it does.

---

*[talpid-ipc/src/client.rs, line 147 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LBucyGDC8KHzQJenhZr:-LBupaabBdoQKYns6v-h:b42f6r4) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2ac4639bfd1de369543c75f7b28f03499466b7d8/talpid-ipc/src/client.rs#L147)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I was confused about this last `_` case. I thought Hmm, but you already covered both `Option` variants. I think it might be easier to parse if this error branch was put before `None` and had the pattern matching `Some(_)` instead of just `_` to make it clear that any other `Some` value is an error.
</blockquote></details>

Done, good idea :)

---

*[talpid-ipc/src/client.rs, line 319 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LBuexIq4UriYoGGIEy2:-LBw-UYM1-M18PbU3SrM:b-ms0gc5) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2ac4639bfd1de369543c75f7b28f03499466b7d8/talpid-ipc/src/client.rs#L319)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

You have defined a number of new error types, but this place uses raw `ErrorKind::Msg`. In libraries I think it's good if one can stick to "real" error variants. Best would be if some errors could be consolidated under the same variant as to not lead to an explosion of error variants, the chain usually tells the underlying reason anyway.
</blockquote></details>

Done. I tried to keep the number of errors not too big, but suggestions for mergers are accepted :)

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by faern** [comment]
Review status: 3 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[talpid-ipc/src/client.rs, line 56 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LBuaqL06zjSDwuecdVm:-LC3xv9_EBY2P68aIyKz:b-4tk5ky) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2ac4639bfd1de369543c75f7b28f03499466b7d8/talpid-ipc/src/client.rs#L56)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

I updated the implementation to have an `unsubscribe` method and I updated the subscription event callback closure to call the `unsubscribe` method when processing the event fails.

However, this was not a simple change. Because the closure runs on a different thread, it was necessary to wrap all of the fields in `Arc<Mutex<_>>`. Then, the problem was that if I simply shared the references between the "main" thread and the "handler" thread, there was the risk of the connection not being closed when all of the user references were dropped. So I decided to make a "weak reference" `WeakWsIpcClient` that works around that limitation by only unsubscribing if the references are still valid.

Things got more complex than expected. I'm not sure the current solution is the best one. However, simply removing all this and letting the user be responsible for unsubscribing seems like a missed opportunity.

I can think of another alternative. I'll see if it works and I'll post it as a new revision if it does.
</blockquote></details>

Whenever I think I have to go towards a `Mutex` I try to look at the problem from far away again. It can usually be avoided. Going into mutexes and weak references is really heavy for just doing a callback to a subscription IMO.

The `WsIpcClient` now has four fields, which are all `Arc<Mutex<.>>`. Either it can be simplified or maybe it's harder to make it simpler because of the design of `ws`, I can't tell as I have not understood everything here yet.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by pinkisemils** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by pinkisemils** [comment]
Reviewed 2 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by pinkisemils** [comment]
Reviewed 1 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by jvff** [comment]
Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[talpid-ipc/src/client.rs, line 56 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LBuaqL06zjSDwuecdVm:-LCUbsCi94dURUK_jMGB:b-hk5i86) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2ac4639bfd1de369543c75f7b28f03499466b7d8/talpid-ipc/src/client.rs#L56)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Whenever I think I have to go towards a `Mutex` I try to look at the problem from far away again. It can usually be avoided. Going into mutexes and weak references is really heavy for just doing a callback to a subscription IMO.

The `WsIpcClient` now has four fields, which are all `Arc<Mutex<.>>`. Either it can be simplified or maybe it's harder to make it simpler because of the design of `ws`, I can't tell as I have not understood everything here yet.
</blockquote></details>

I'm struggling to find a solution for this problem. The root cause is that there are two execution threads, which can be thought as the "sender" (which owns the `WsIpcClient`) and the "receiver" (which is the `ws` WebSocket handler thread). Every time the "sender" sends a request, it registers the message ID in a hash-map shared with the "receiver", so that it knows that a given message is a response to a request. That's working fine, the problem is when a message is received and it has to send another message. In this specific case, a notification message is received, an error occurs and it has to send an `unsubscribe` request.

The previous solution shared only the hash-map between the threads when sending a message, because the "receiver" thread only needs to know what requests are waiting for a response. However, everything must be shared when the "receiver" needs to send a message. I don't want to simply use a normal `Arc` for this otherwise the "sender" thread, once terminated, won't close the connection as long as the "receiver" thread is still active (because the receiver is sharing the connection handle). That's why in the previous solution the "receiver" thread has a `Weak` reference. The connection gets closed successfully if the "sender" thread finishes, and the "receiver" thread can send messages as long as the "sender" thread is still alive.

However, there are many factors in the previous implementation that leads to increased complexity. In this new approach, the complexity is still there (a lot of changes were necessary), but I hope I managed to better organize things. Now there's a third thread that handles sending and receiving, and the original two threads just communicate with the new thread using `mpsc` channels. I believe I moved the complexity to different areas, so now in terms of architecture there are three threads, but all "sharing" has been contained (or rather removed) to be in one single manager/handler thread. Therefore the logic in the third thread (which is by itself not so simple) is not obstructed by code to organize how data is shared.

I'm still hoping a simpler solution emerges. I have a feeling that when we change things to `async` it will be better organized, but the problem is that it might appear to be better to start with changes to `ws`, which is probably not a good idea right now. Anyway, any feedback is appreciated. This solution and the previous one work, so we might want to see which one is more palatable, or if there are suggestions for different approaches.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by faern** [comment]
Reviewed 1 of 2 files at r2.
Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[talpid-ipc/src/client.rs, line 56 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LBuaqL06zjSDwuecdVm:-LCbyY9N40RVaWwzoBr5:b-mw3ycd) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2ac4639bfd1de369543c75f7b28f03499466b7d8/talpid-ipc/src/client.rs#L56)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

I'm struggling to find a solution for this problem. The root cause is that there are two execution threads, which can be thought as the "sender" (which owns the `WsIpcClient`) and the "receiver" (which is the `ws` WebSocket handler thread). Every time the "sender" sends a request, it registers the message ID in a hash-map shared with the "receiver", so that it knows that a given message is a response to a request. That's working fine, the problem is when a message is received and it has to send another message. In this specific case, a notification message is received, an error occurs and it has to send an `unsubscribe` request.

The previous solution shared only the hash-map between the threads when sending a message, because the "receiver" thread only needs to know what requests are waiting for a response. However, everything must be shared when the "receiver" needs to send a message. I don't want to simply use a normal `Arc` for this otherwise the "sender" thread, once terminated, won't close the connection as long as the "receiver" thread is still active (because the receiver is sharing the connection handle). That's why in the previous solution the "receiver" thread has a `Weak` reference. The connection gets closed successfully if the "sender" thread finishes, and the "receiver" thread can send messages as long as the "sender" thread is still alive.

However, there are many factors in the previous implementation that leads to increased complexity. In this new approach, the complexity is still there (a lot of changes were necessary), but I hope I managed to better organize things. Now there's a third thread that handles sending and receiving, and the original two threads just communicate with the new thread using `mpsc` channels. I believe I moved the complexity to different areas, so now in terms of architecture there are three threads, but all "sharing" has been contained (or rather removed) to be in one single manager/handler thread. Therefore the logic in the third thread (which is by itself not so simple) is not obstructed by code to organize how data is shared.

I'm still hoping a simpler solution emerges. I have a feeling that when we change things to `async` it will be better organized, but the problem is that it might appear to be better to start with changes to `ws`, which is probably not a good idea right now. Anyway, any feedback is appreciated. This solution and the previous one work, so we might want to see which one is more palatable, or if there are suggestions for different approaches.
</blockquote></details>

I have not read the new solution yet. But just a comment about the earlier shared `HashMap` for getting the message ID from the sender to the receiver: It could have been an `mpsc` channel that the sender send new IDs in to the receiver. And when the receiver gets a message it can first check a `HashMap` (that it owns, without any mutexes) and if not present it can check the channel and any ID coming out of the channel that is not the current one is just stacked in the map for later. That way it uses locally owned collections and communicate with non blocking message passing.

Now I will go on to read the new solution :)

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by faern** [comment]
Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[talpid-ipc/src/client.rs, line 56 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LBuaqL06zjSDwuecdVm:-LCbzPQ35xGS2Ji-iljj:b-cgysfa) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2ac4639bfd1de369543c75f7b28f03499466b7d8/talpid-ipc/src/client.rs#L56)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I have not read the new solution yet. But just a comment about the earlier shared `HashMap` for getting the message ID from the sender to the receiver: It could have been an `mpsc` channel that the sender send new IDs in to the receiver. And when the receiver gets a message it can first check a `HashMap` (that it owns, without any mutexes) and if not present it can check the channel and any ID coming out of the channel that is not the current one is just stacked in the map for later. That way it uses locally owned collections and communicate with non blocking message passing.

Now I will go on to read the new solution :)
</blockquote></details>

Regarding the future and `async` and all. Yeah, I'm starting to think we can more or less just go with any solution that works for now. It does not have to be optimal in neither performance nor code beauty :). In a not too distant future we will hopefully be able to replace `ws` with either just `hyper` or some other more ergonomic WebSockets library. Then we'll have to basically rewrite from scratch anyway.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by faern** [comment]
Reviewed 1 of 4 files at r3.
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[talpid-ipc/src/client.rs, line 130 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LCc-BvgEG9Fxb_GM_vb:-LCc-BvgEG9Fxb_GM_vc:b-holzn9) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d3838430d3106fcc46a49426a26967139d18a8b8/talpid-ipc/src/client.rs#L130)):*
> ```Rust
>     Call {
>         method: String,
>         arguments: JsonValue,
> ```

It's not important now. But just want to highlight this: Doing intermediate storage as a `JsonValue` requires serializing in two steps. If one can go directly from `T` to the final string representation that is faster. But yeah, performance is not a priority at all at this point.

---

*[talpid-ipc/src/client.rs, line 348 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LCc2xDo6EBm-pZpu1rv:-LCc2xDo6EBm-pZpu1rw:b722ddu) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d3838430d3106fcc46a49426a26967139d18a8b8/talpid-ipc/src/client.rs#L348)):*
> ```Rust
>         M: From<V> + Send + 'static,
>     {
>         let raw_subscription_id = self.call(&format!("{}_subscribe", event), &[] as &[u8; 0])?;
> ```

That a subscribe method ends in `_subscribe` is not something enforced by the protocol. It's just something we have applied to all our mullvad-daemon methods. If this is kept we kind of make this `talpid-ipc` (talpid == mullvad agnostic) type aware of, and bound to, our method naming style defined at the mullvad specific layer.

I think it would be easy to let the caller append `_subscribe` themselves if they want to talk to an API with that naming standard? And it would make this code more generic as well.

---

*[talpid-ipc/src/client.rs, line 508 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LCc5THg7kTw9kwgNEmb:-LCc5THg7kTw9kwgNEmc:b-3lc7hh) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d3838430d3106fcc46a49426a26967139d18a8b8/talpid-ipc/src/client.rs#L508)):*
> ```Rust
>                 request.send_response(result);
>             } else {
>                 self.active_request = Some(request);
> ```

I don't fully understand the `active_request` thing. So this type can only handle one request at a time. And the response for that has to arrive before the next is sent? What if we get a notification while we wait for a response to a call? Does that not end up here and cause the warning/error?

---

*[talpid-ipc/src/client.rs, line 570 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LCc3uik0dnSxae9WWxc:-LCc3uik0dnSxae9WWxd:bjizs52) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d3838430d3106fcc46a49426a26967139d18a8b8/talpid-ipc/src/client.rs#L570)):*
> ```Rust
>     let value: V = serde_json::from_value(json_value)
>         .chain_err(|| ErrorKind::DeserializeSubscriptionEvent(event_name.clone()))?;
>     let message = value.into();
> ```

I personally try to avoid `.into()` since it's a bit hard to follow along what type it becomes. Would you be comfortable with `M::from(value)`?

---

*[talpid-ipc/src/lib.rs, line 9 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LCc6Ft54wLNq4T65Gj9:-LCc6Ft54wLNq4T65GjA:b-f1mr7z) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d3838430d3106fcc46a49426a26967139d18a8b8/talpid-ipc/src/lib.rs#L9)):*
> ```Rust
> //! the License, or (at your option) any later version.
> 
> #![recursion_limit = "128"]
> ```

Which part of the new code requires deeper recursion while compiling?

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by pinkisemils** [comment]
Reviewed 2 of 4 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by jvff** [comment]
Review status: 3 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[talpid-ipc/src/client.rs, line 56 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LBuaqL06zjSDwuecdVm:-LCcN9g2BW8RqUAWv-HE:b-9ysti6) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2ac4639bfd1de369543c75f7b28f03499466b7d8/talpid-ipc/src/client.rs#L56)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

You have done a massive effort to make this work and get the pubsub feature into this crate. That alone is amazing. I did not mean to say the work was in vain by saying it does not matter if the code is perfect for now, of course a nice solution is better than a not so nice solution :) And I suspect it will be a couple of months at least before we replace any of this.
</blockquote></details>

Don't worry, we need a working solution, but if a problem appears in the code before we manage to rewrite it, its best if it's easier to find the cause. So we should try to get the best solution possible, but I think this case the results we're getting for the effort spent is starting to diminish, so I think we can just go with this for now.

---

*[talpid-ipc/src/client.rs, line 348 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LCc2xDo6EBm-pZpu1rv:-LCcc1pA3vVS_8r6nh_m:bn03m0a) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d3838430d3106fcc46a49426a26967139d18a8b8/talpid-ipc/src/client.rs#L348)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

That a subscribe method ends in `_subscribe` is not something enforced by the protocol. It's just something we have applied to all our mullvad-daemon methods. If this is kept we kind of make this `talpid-ipc` (talpid == mullvad agnostic) type aware of, and bound to, our method naming style defined at the mullvad specific layer.

I think it would be easy to let the caller append `_subscribe` themselves if they want to talk to an API with that naming standard? And it would make this code more generic as well.
</blockquote></details>

Done, but now this requires the `subscribe` method to have `subscribe_method` and `unsubscribe_method` parameters. The `unsubscribe_method` parameter is used in order to unsubscribe when an error occurs or the `mpsc` channel is closed.

---

*[talpid-ipc/src/client.rs, line 508 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LCc5THg7kTw9kwgNEmb:-LCcYV34EcSWQ_B9EUpR:bfuo45y) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d3838430d3106fcc46a49426a26967139d18a8b8/talpid-ipc/src/client.rs#L508)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I don't fully understand the `active_request` thing. So this type can only handle one request at a time. And the response for that has to arrive before the next is sent? What if we get a notification while we wait for a response to a call? Does that not end up here and cause the warning/error?
</blockquote></details>

Yes, you are correct regarding the serialization of requests. Only one is handled at a time, and the response must be received before another request is sent. I believe this is fine for now because `WsIpcClient` is currently synchronous. However, since it only contains a `mpsc::Sender` to the connection handler, it currently can't be cloned, otherwise it would allow concurrent requests.

It's not hard to replace the `Option<ActiveRequest>` with a `HashMap<i64, ActiveRequest>`, so it's possible to support multiple in-flight requests. However, I thought this way would be simpler for now. I can change it if necessary. Doing so would allow a `#[derive(Clone)]` for `WsIpcClient` to share handles to the same connection.

Regarding conflict with notifications, there is none because the code paths are separate. Requests and responses (i.e., messages received with a request ID) use the `active_request` field, while subscriptions and notifications (i.e., messages received without a request ID) use the `active_subscriptions` `HashMap`.

---

*[talpid-ipc/src/client.rs, line 570 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LCc3uik0dnSxae9WWxc:-LCcPwx8BX8bErcB-W1X:b5s2vcc) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d3838430d3106fcc46a49426a26967139d18a8b8/talpid-ipc/src/client.rs#L570)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I personally try to avoid `.into()` since it's a bit hard to follow along what type it becomes. Would you be comfortable with `M::from(value)`?
</blockquote></details>

Sure, no problem. Done.

---

*[talpid-ipc/src/lib.rs, line 9 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LCc6Ft54wLNq4T65Gj9:-LCcO7vdEhRG8VpzgzaG:b1u5fgd) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d3838430d3106fcc46a49426a26967139d18a8b8/talpid-ipc/src/lib.rs#L9)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Which part of the new code requires deeper recursion while compiling?
</blockquote></details>

`error-chain` :/ Currently the `ErrorKind` list has grown too large.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by pinkisemils** [comment]
Reviewed 1 of 4 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by faern** [comment]
Reviewed 1 of 4 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[talpid-ipc/src/client.rs, line 508 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LCc5THg7kTw9kwgNEmb:-LCdaK8fEQryVQWz0E_H:b-5559zl) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d3838430d3106fcc46a49426a26967139d18a8b8/talpid-ipc/src/client.rs#L508)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Yes, you are correct regarding the serialization of requests. Only one is handled at a time, and the response must be received before another request is sent. I believe this is fine for now because `WsIpcClient` is currently synchronous. However, since it only contains a `mpsc::Sender` to the connection handler, it currently can't be cloned, otherwise it would allow concurrent requests.

It's not hard to replace the `Option<ActiveRequest>` with a `HashMap<i64, ActiveRequest>`, so it's possible to support multiple in-flight requests. However, I thought this way would be simpler for now. I can change it if necessary. Doing so would allow a `#[derive(Clone)]` for `WsIpcClient` to share handles to the same connection.

Regarding conflict with notifications, there is none because the code paths are separate. Requests and responses (i.e., messages received with a request ID) use the `active_request` field, while subscriptions and notifications (i.e., messages received without a request ID) use the `active_subscriptions` `HashMap`.
</blockquote></details>

Cool. Thanks for the explanation. Yes, only handling one request at a time is perfectly fine for now :ok_hand: 

---

*[talpid-ipc/src/lib.rs, line 9 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159#-LCc6Ft54wLNq4T65Gj9:-LCdaRD24KpC8g6o_ea7:b-2t18yn) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d3838430d3106fcc46a49426a26967139d18a8b8/talpid-ipc/src/lib.rs#L9)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

`error-chain` :/ Currently the `ErrorKind` list has grown too large.
</blockquote></details>

I see. Yeah this is probably the longest list of error chain errors we have.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/159)*

**Comment by jvff** [comment]
This should be fixed by using the latest version of the `rustc` compiler (perhaps by running `rustup update`?).

Error `E0658` means that [an unstable feature was used](https://doc.rust-lang.org/error-index.html#E0658), but since version 1.0.26, the specific [feature in question](https://blog.rust-lang.org/2018/05/10/Rust-1.26.html) (in the "Nicer `match` bindings" section) has been stabilized, and the `&` can be omitted.

Thank you very much for the report, and I hope this helps :)
